core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java

```diff
@@ -1223,4 +1224,46 @@ default MaterializedViewFreshness getMaterializedViewFreshness(ConnectorSession
     {
         return Optional.empty();
     }
+
+    /**
+     * Return a list of table layouts that satisfy the given constraint, for a given table version
+     */
+    default List<ConnectorTableLayoutResult> getTableLayouts(
+            ConnectorSession session,
+            ConnectorTableHandle table,
+            Constraint constraint,
+            Optional<Set<ColumnHandle>> desiredColumns,
+            Optional<ConnectorVersionHandle> version)
+    {
+        throw new TrinoException(NOT_SUPPORTED, "This connector does not support versioning on tables");
+    }
+
+    default ConnectorTableLayout getTableLayout(ConnectorSession session, ConnectorTableLayoutHandle handle, ConnectorVersionHandle version)
+    {
+        throw new TrinoException(NOT_SUPPORTED, "This connector does not support versioning on tables");
+    }
+
+    /**
+     * Returns a table handle representing a versioned table. A versioned table differs by having an additional specifier for version.
+     */
+    default ConnectorTableHandle getTableHandle(ConnectorSession session, SchemaTableName tableName, Optional<ConnectorVersionHandle> version)
+    {
+        throw new TrinoException(NOT_SUPPORTED, "This connector does not support versioned tables");
+    }
+
+    /**
+     * Returns whether a specified version type is valid for a given travel type and table name
+     */
+    default boolean isValidVersionType(ConnectorSession session, SchemaTableName tableName, TravelType travelType, Type versioning)
```

Comment:
It's the data type of versioning literal being used for the query period. Most typically this would be TIMESTAMP WITH TIME ZONE for temporal travel and STRING or INT for version travel, but leaving this as a connector implementation detail since support varies. 

The data types that seem to be commonly supported for various data lake solutions seem to be:
Temporal Travel: Timestamp, Timestamp with time zone, BigInt, String
Version Travel: String, Int, BigInt

Reply:
Thanks for the clarification.

For temporal type, I think we should stick to TIMESTAMP WITH TIME ZONE. First, it's what the spec expects. Second, any conversion or reinterpretation/mapping to other types should be done inside the connector (e.g., if a connector uses unix time or something similar).

On another note, this API allows varying the allowed type on a per-table and travel type basis. For version travel, we should consider whether the type should vary on a table by table basis or whether it should be a connector-wide option. Can you think of any cases where the former is necessary?

Reply:
Regarding the second point, whether or not we should allow type on a per travel-type/table basis: the API assumes that valid versions can change based on characteristics of individual tables, which is why the table name is also passed in. One example would be if a database allowed system versioning, but the setting was turned off for a particular table.

The by-product of checking whether a particular version type is handled by a table supports a specific type is that it also determines if temporal travel, version travel, or some combination of them is supported. Rather than creating separate SPIs for the different travel types, I thought it was a clean and extensible abstraction. 

As far as support for temporal travel with types other than TIMESTAMP WITH TIME ZONE, I only have access to the free ISO 19075 pdf from 2015 that details system versioning, but I don't see the restriction to TS with TZ, the examples seem to be in TS and Date format. I'm still a bit hesitant to add in the restriction at the engine level because it seems like there are still valid use cases for isValidVersionType with temporal travel even restricting the type to explicitly TS with TZ (whether or not temporal travel is supported at all for a particular table, or to enforce a certain level of precision TS(6) vs TS(9)).

Reply:
You're right. The spec says it must be "DATE or a timestamp type" (i.e., `timestamp with time zone` or `timestamp without time zone`)